From f6a2860298bce0b6e2c8987d804974abc8d8ba47 Mon Sep 17 00:00:00 2001
From: Robin Gong <b38343@freescale.com>
Date: Mon, 17 Aug 2015 17:29:06 +0800
Subject: [PATCH 1587/5242] MLK-11358-6: dma: imx-sdma: reload context once
 sdma channel terminated

commit  a3053e7462da2f8b3d58259aff2dfba7aa456920 from
https://source.codeaurora.org/external/imx/linux-imx.git

Some driver may call dmaengine_terminate_all firstly, and then start next by
calling dmaengine_prep_* without dmaengine_slave_config. In this case sdma
transfer failed since no context loaded, take this case in this patch.

Signed-off-by: Robin Gong <b38343@freescale.com>
(cherry picked from commit 51ff948df543dc273fefeb86608e7d6d28ca8090)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/dma/imx-sdma.c |   12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/dma/imx-sdma.c b/drivers/dma/imx-sdma.c
index 3a91e88..bf45b74 100644
--- a/drivers/dma/imx-sdma.c
+++ b/drivers/dma/imx-sdma.c
@@ -344,6 +344,7 @@ struct sdma_channel {
 	struct imx_dma_data		data;
 	unsigned int			chn_count;
 	unsigned int			chn_real_count;
+	bool				context_loaded;
 	u32				bd_size_sum;
 	bool				enabled;
 	bool				src_dualfifo;
@@ -986,6 +987,9 @@ static int sdma_load_context(struct sdma_channel *sdmac)
 	int ret;
 	unsigned long flags;
 
+	if (sdmac->context_loaded)
+		return 0;
+
 	if (sdmac->direction == DMA_DEV_TO_MEM)
 		load_address = sdmac->pc_from_device;
 	else if (sdmac->direction == DMA_DEV_TO_DEV)
@@ -1033,6 +1037,8 @@ static int sdma_load_context(struct sdma_channel *sdmac)
 
 	spin_unlock_irqrestore(&sdma->channel_0_lock, flags);
 
+	sdmac->context_loaded = true;
+
 	return ret;
 }
 
@@ -1363,6 +1369,7 @@ static int sdma_terminate_all(struct dma_chan *chan)
 	spin_unlock_irqrestore(&sdmac->vc.lock, flags);
 	vchan_dma_desc_free_list(&sdmac->vc, &head);
 	sdma_disable_channel(chan);
+	sdmac->context_loaded = false;
 
 	return 0;
 }
@@ -1477,7 +1484,11 @@ static struct sdma_desc *sdma_transfer_init(struct sdma_channel *sdmac,
 	if (sdma_alloc_bd(desc))
 		goto err_desc_out;
 
+	if (sdma_load_context(sdmac))
+		goto err_desc_out;
+
 	return desc;
+
 err_desc_out:
 	kfree(desc);
 err_out:
@@ -2235,6 +2246,7 @@ static int sdma_probe(struct platform_device *pdev)
 		struct sdma_channel *sdmac = &sdma->channel[i];
 
 		sdmac->sdma = sdma;
+		sdmac->context_loaded = false;
 		sdmac->channel = i;
 		sdmac->status = DMA_IN_PROGRESS;
 		sdmac->vc.desc_free = sdma_desc_free;
-- 
1.7.9.5

